TCD-2899: Backtrace Language Integrations (#2148)
saucelabs/sauce-docs: docs/error-reporting/language-integrations/mixed.md

---
id: mixed
title: Mixed Call Stacks with C# and C++
sidebar_label: C# and C++
description: Use C# and C++ in your Backtrace project.
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import useBaseUrl from '@docusaurus/useBaseUrl';

> C# is a common language used to create cross-platform applications and user interfaces. In many cases, C# and C++ are used together to create more sophisticated applications. In some scenarios, a C# application will make calls to functions within managed or unmanaged C++ libraries. In others, a C++ application may act as the shell for the app, and cross-platform C# UI code is used within it. In either case, engineers require detailed crash and exception report data from both the C# and C++ applications to effectively debug and identify the root cause of any faulty code.

Author: ```suggestion
C# is a common language for creating cross-platform applications and user interfaces. In many cases, C# and C++ are used together to create more sophisticated applications. In some scenarios, a C# application will make calls to functions in managed or unmanaged C++ libraries. In others, a C++ application may act as the shell for the app, and cross-platform C# UI code is used in it. In either case, engineers require detailed crash and exception report data from both the C# and C++ applications to effectively debug and identify the root cause of any faulty code.
```

1. Select **+**.

:::

## Description And UML Activity Diagram of Flow Control Between C# And C++

In scenarios where a C# or C++ application calls unmanaged C++ code or a C# method, and that code throws an exception, the crash reporting platform needs to ensure that call stacks and other relevant data from both portions of code are available for analysis.

The figure below presents UML Activity diagrams depicting how the code is executed and what information is provided to identify the calling and failing code.

<img src={useBaseUrl('/img/error-reporting/language-integrations/5e601be3ca0c9.png')} alt="UML Activity diagrams"/>

The C# environment requires a callback method, invoked by the C++ environment, to allow the developer to send additional data from the C# application state together with C++ report generated by Crashpad. The system also requires that an additional attribute - `process_id` - is added to the Crashpad and Backtrace C# Report to identify both parts of the call stack.

The remainder of this document will discuss how you can configure your application to report exceptions and crashes to your Backtrace instance. Then, you will find sample code showing how to.

> ## How to Set up Crash And Exception Reporting in C# And C++

Author: ```suggestion
## How to Setup Crash And Exception Reporting in C# And C++
```

In our example, we want to set up a callback function. To prepare the callback, we create a delegate method with different types of parameters: `IntPtr`, `int`, and `int`. In our case, we use the callback function to invoke C# code before the Crashpad `DumpWithoutCrash` method. You can check the delegate declaration and usage below:

```csharp
delegate void LogBuffer(IntPtr buf, int len, int flags);

private LogBuffer logger;
private BacktraceClient client;

public void OnLogMessage(IntPtr _buf, int len, int flags)
{
    if (len == 0 || _buf == IntPtr.Zero)
    {
        Console.WriteLine("Buffer or len is empty.");
    }
    byte[] managedArray = new byte[len - 1];
    Marshal.Copy(_buf, managedArray, 0, len - 1);
    Console.WriteLine(managedArray);

    string result = Encoding.UTF8.GetString(managedArray);
    Console.WriteLine(result);
    var report = new BacktraceReport(
        message: "Event caught by C#",
        attributes: new Dictionary<string, object>() {
            { "process.id", result }
        }
    );
    Task.WaitAll(client.SendAsync(report));
}

public void ExecuteTasks()
{
    logger = OnLogMessage;
    CrashApp(logger);
}
```

### Execute C++ Methods

> If you want to execute C++ library code from C#, use the following code to export your method:

Author: ```suggestion
If you want to run C++ library code from C#, use the following code to export your method:
```

```csharp
extern "C" _declspec(dllexport) methodType MethodName(method args...) {}
```

The following is a more detailed implementation that allows you to pass a callback function to C#. (Check the [Invoking C++ Methods from C# Code](#invoking-c-methods-from-c-code) section above to read more about callback functions.)

```csharp
extern "C" _declspec(dllexport) void CrashApp(LogBuffer g_logger) {
    LogBuffer buffer = g_logger;
    BacktraceClient client = BacktraceClient::BacktraceClient();
    if (buffer != IntPtr.Zero) {
        client.SetupLogger(buffer);
    }
    client.Crash();
}
```

> ### C++ Calling into C#

Author: ```suggestion
### C++ Calling Into C#
```
